File: Data Structures & Algorithms/subsets-ii/submission-3.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> subsets = new ArrayList<>();
        List<Integer> currSet = new ArrayList<>();
        helper(0, nums, currSet, subsets);
        return subsets;
    }

    private void helper(int i, int[] nums, List<Integer> currSet, List<List<Integer>> subsets) {
        if (i >= nums.length) {
            subsets.add(new ArrayList<>(currSet));
            return;
        }

        currSet.add(nums[i]);
        helper(i + 1, nums, currSet, subsets);

        currSet.remove(currSet.size() - 1);
        // we do not want to try the element at index (currSet.size() - 1)
        // but we ALSO DO NOT WANT TO TRY IT'S DUPLICATES. SO we have to
        // skip and make sure that, right after removing X, we just end up
        // adding and exploring with it's duplicate. That defies the whole
        // point.
        while ((i + 1 < nums.length) && (nums[i] == nums[i + 1])) {
            i++;
        }
        helper (i + 1, nums, currSet, subsets);
    }
}

File: Data Structures & Algorithms/remove-duplicates-from-sorted-array/submission-3.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        count = 0
        write = None
        idx = 0
        for i in range(len(nums)):
            if write:
                nums[idx] = nums[i]
                write = False
            if i + 1 < len(nums):
                if nums[i] == nums[i + 1] and not write:
                    count += 1
                    write = True
                    idx = i + 1
        return len(nums) - count


"""
- Remember that the array is sorted. So all the duplicates would be right next to
each other.
- Remove all the duplicates in-place. After removing the duplicates, return the
number of unique elements, denoted as k. Such that the first k elements of nums
contain the unique elements.
- First k elements must contain all the unique elements.
- Have a write index which stores the index to which you can write. But write what?
- Write the next available element.
- So if we recently get an index to which we can write, we introduce some sort of
boolean flag which would tell us, that we are ready to write.
"""